File: backend/nfe/src/model/nfe_item_rastreado_model.py
from src import db


class NfeItemRastreadoModel(db.Model):
    __tablename__ = 'nfe_item_rastreado'

    id = db.Column(db.Integer, primary_key=True)
    numero_lote = db.Column(db.String(20))
    quantidade_itens = db.Column(db.Float)
    data_fabricacao = db.Column(db.DateTime)
    data_validade = db.Column(db.DateTime)
    codigo_agregacao = db.Column(db.String(20))
    id_nfe_detalhe = db.Column(db.Integer, db.ForeignKey('nfe_detalhe.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_detalhe = data.get('idNfeDetalhe')
        self.numero_lote = data.get('numeroLote')
        self.quantidade_itens = data.get('quantidadeItens')
        self.data_fabricacao = data.get('dataFabricacao')
        self.data_validade = data.get('dataValidade')
        self.codigo_agregacao = data.get('codigoAgregacao')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeDetalhe': self.id_nfe_detalhe,
            'numeroLote': self.numero_lote,
            'quantidadeItens': self.quantidade_itens,
            'dataFabricacao': self.data_fabricacao.isoformat(),
            'dataValidade': self.data_validade.isoformat(),
            'codigoAgregacao': self.codigo_agregacao,
        }

File: frontend/nfe/lib/app/data/model/nfe_item_rastreado_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/infra/infra_imports.dart';
import 'package:intl/intl.dart';

class NfeItemRastreadoModel {
	int? id;
	int? idNfeDetalhe;
	String? numeroLote;
	double? quantidadeItens;
	DateTime? dataFabricacao;
	DateTime? dataValidade;
	String? codigoAgregacao;

	NfeItemRastreadoModel({
		this.id,
		this.idNfeDetalhe,
		this.numeroLote,
		this.quantidadeItens,
		this.dataFabricacao,
		this.dataValidade,
		this.codigoAgregacao,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero_lote',
		'quantidade_itens',
		'data_fabricacao',
		'data_validade',
		'codigo_agregacao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero Lote',
		'Quantidade Itens',
		'Data Fabricacao',
		'Data Validade',
		'Codigo Agregacao',
	];

	NfeItemRastreadoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeDetalhe = jsonData['idNfeDetalhe'];
		numeroLote = jsonData['numeroLote'];
		quantidadeItens = jsonData['quantidadeItens']?.toDouble();
		dataFabricacao = jsonData['dataFabricacao'] != null ? DateTime.tryParse(jsonData['dataFabricacao']) : null;
		dataValidade = jsonData['dataValidade'] != null ? DateTime.tryParse(jsonData['dataValidade']) : null;
		codigoAgregacao = jsonData['codigoAgregacao'];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeDetalhe'] = idNfeDetalhe != 0 ? idNfeDetalhe : null;
		jsonData['numeroLote'] = numeroLote;
		jsonData['quantidadeItens'] = quantidadeItens;
		jsonData['dataFabricacao'] = dataFabricacao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataFabricacao!) : null;
		jsonData['dataValidade'] = dataValidade != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataValidade!) : null;
		jsonData['codigoAgregacao'] = codigoAgregacao;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeDetalhe = plutoRow.cells['idNfeDetalhe']?.value;
		numeroLote = plutoRow.cells['numeroLote']?.value;
		quantidadeItens = plutoRow.cells['quantidadeItens']?.value?.toDouble();
		dataFabricacao = Util.stringToDate(plutoRow.cells['dataFabricacao']?.value);
		dataValidade = Util.stringToDate(plutoRow.cells['dataValidade']?.value);
		codigoAgregacao = plutoRow.cells['codigoAgregacao']?.value;
	}	

	NfeItemRastreadoModel clone() {
		return NfeItemRastreadoModel(
			id: id,
			idNfeDetalhe: idNfeDetalhe,
			numeroLote: numeroLote,
			quantidadeItens: quantidadeItens,
			dataFabricacao: dataFabricacao,
			dataValidade: dataValidade,
			codigoAgregacao: codigoAgregacao,
		);			
	}

	
}